### PR TITLE
fix(router): avoid dense elements allocation for numeric URL keys

### LINKED ISSUE #70716: Angular Router numeric matrix parameters can exhaust memory during SSR
### Which @angular/* package(s) are the source of the bug?

router

### Is this a regression?

No

### Description

Angular Router stores matrix parameters in regular JavaScript objects.

Numeric parameter names such as:

```text
/a;990;2522
```

behave very differently in V8 compared with normal string keys.

For example, these two paths have the same length and contain the same number of matrix parameters:

```text
/a;990;2522
/a;aaa;bbbb
```

but the numeric version uses significantly more memory.

During Angular SSR, this difference becomes large enough that a relatively small burst of concurrent requests can exhaust the Node.js heap.

I tested this through Nginx using request sizes that still fit within its normal 8 KiB request-line limit:

| V8 old space | Concurrent requests | Segments/request | Request path size |
| ---: | ---: | ---: | ---: |
| 256 MiB | 12 | 742 | 8,162 B |
| 512 MiB | 22 | 742 | 8,162 B |

With a 256 MiB heap, 12 concurrent requests using the numeric version consistently terminate the SSR worker with:

```text
JavaScript heap out of memory
```

The same-size control using normal string keys completes successfully and Angular remains healthy.

The reproduction briefly keeps requests active for 500 ms so that their SSR work overlaps consistently. This is only to make the memory difference easy to reproduce.

### Please provide a link to a minimal reproduction of the bug

https://github.com/SkyZeroZx/angular-router-ssr-oom-lab

### PR DESCRIPTION
**Pull request by SkyZeroZx**
Keep numeric URL keys in V8 dictionary elements during parsing and copying to avoid disproportionate backing-store allocation.

Preserve parameter values, outlet names, property descriptors and merge precedence across redirects, snapshots and component input binding.

Fixes #70716

---

In  [#70716](https://github.com/angular/angular/issues/70716) shows that numeric matrix-parameter names such as:

```text
/a;990;2522
```

can cause disproportionate V8 backing-store allocation during Angular SSR.

A minimal POC through Nginx with an 8,162-byte request path, still within its normal _8 KB_ request line limit. 

The following table uses nginx with 8kb, using the default configuration :

| V8 old space | Concurrent requests | Result |
| ---: | ---: | --- |
| 256 MiB | 12 | V8 heap OOM |
| 512 MiB | 22 | V8 heap OOM |

NOTE : Other common deployments allow larger request lines; e.g  AWS ALB allows use _16 KB_.

The requests are briefly overlapped during SSR to make the memory behavior reproducible. A same-size path using ordinary string keys completes normally.

See: https://github.com/SkyZeroZx/angular-router-ssr-oom-lab

### CODE REVIEW

File: packages/router/src/utils/collection.ts [hunk 1]
```diff
@@ -91,3 +91,63 @@ export function wrapIntoPromise<T>(value: T | Promise<T> | Observable<T>): Promi
...
+  ) {
+    Object.defineProperty(target, SLOW_ELEMENTS_SENTINEL, {value: 0, configurable: true});
+    delete target[SLOW_ELEMENTS_SENTINEL];
+  }
+}
+
+/** Assigns a URL-derived key without sparse-index allocation in V8. */
+export function setUrlDerivedKey<T>(target: {[key: string]: T}, key: string, value: T): void {
+  preventDenseElements(target, key);
+  target[key] = value;
```
Comment (arturovt) [inline packages/router/src/utils/collection.ts]: just a Q: should we use Object.defineProperty insttead to avoid proto pollution?

Comment (SkyZeroZx) [reply packages/router/src/utils/collection.ts]: Good catch , done

**Review by atscott — CHANGES_REQUESTED** [review]
AGENT: This PR is much broader and more complex than necessary because it replaces `{...}` object spreads across route matching, snapshots, and `<router-outlet>` input binding (`mergeUrlDerivedKeys` / `defineUrlDerivedKey`):

1. **Scope of the amplification (`UrlTree` parsing vs. Route Snapshots):**
   - A single object with keys `990` and `2522` only allocates ~36 KB in V8. The 25.5 MB heap spike happens because a single 8 KB URL can contain **742 `UrlSegment`s**, creating 742 separate `segment.parameters` objects during `UrlParser` parsing (or hundreds of sibling `UrlSegmentGroup.children` maps).
   - By contrast, `UrlTree.queryParams` exists **once per URL**, and `ActivatedRouteSnapshot.params` / `RoutedComponentInputBinder` `data` only exist **once per matched `Route` in the application's `Routes` configuration** (not once per URL segment—even a `{ path: '**' }` wildcard route only merges `res.consumed[res.consumed.length - 1].parameters`, i.e., a single segment).
   - Replacing native `{...}` spreads across `recognize.ts`, `router.ts`, `router_state.ts`, `router_outlet.ts`, and `config_matching.ts` with `Reflect.ownKeys()` + `Object.prototype.propertyIsEnumerable.call()` + `Object.defineProperty()` adds runtime overhead on every navigation and increases bundle size (5 new golden symbols), while saving at most ~36 KB on 1–5 objects.
   - Moreover, because V8's dictionary-elements representation never survives `{...}` spread in application code or third-party libraries anyway, trying to maintain dictionary mode across all `{...}` spreads is an unmaintainable invariant.

2. **Please scope this PR down to just `setUrlDerivedKey`:**
   - Remove `mergeUrlDerivedKeys` and `defineUrlDerivedKey` completely.
   - Revert `packages/router/src/recognize.ts`, `packages/router/src/router.ts`, `packages/router/src/router_state.ts`, `packages/router/src/directives/router_outlet.ts`, and `packages/router/src/utils/config_matching.ts` (and their corresponding spec/golden symbol changes) so they continue using native `{...}` spread.
   - Keep `setUrlDerivedKey` only where `UrlTree` parameter/outlet records are populated (`url_tree.ts`, and the segment group clones in `apply_redirects.ts` / `create_url_tree.ts`).

3. **Simplify `setUrlDerivedKey` in `packages/router/src/utils/collection.ts`:**
   - Indices `< 32` allocate a tiny `FixedArray` (at most ~62 slots, comparable in size to a `NumberDictionary`), so we only need to transition when `index >= 32`.
   - Using `Object.defineProperty` on plain `{}` parameter maps alters `__proto__` key behavior (defining an own `'__proto__'` property on `params`/`queryParams` where `params['__proto__'] = 'str'` was previously a no-op). Standard property assignment `target[key] = value` after the sentinel transition is sufficient (`UrlSegmentGroup.children` maps that need `__proto__` protection already use `Object.create(null)`):

   ```ts
   const SLOW_ELEMENTS_SENTINEL = 0x40000000;

   export function setUrlDerivedKey<T>(target: {[key: string]: T}, key: string, value: T): void {
     const index = Number(key) >>> 0;
     if (
       index >= 32 &&
       index !== 0xffffffff &&
       String(index) === key &&
       !Object.hasOwn(target, SLOW_ELEMENTS_SENTINEL)
     ) {
       target[SLOW_ELEMENTS_SENTINEL] = 0 as unknown as T;
       delete target[SLOW_ELEMENTS_SENTINEL];
     }
     target[key] = value;
   }
   ```

**Review by atscott — CHANGES_REQUESTED** [review]
AGENT: Thanks for trimming this down. Looking at the remaining files, we can simplify this further and keep it entirely local to `packages/router/src/url_tree.ts`:

1. **Revert `apply_redirects.ts`, `create_url_tree.ts`, and `utils/collection.ts` (and their specs):**
   - In `apply_redirects.ts`, `createSegmentGroup` walks `this.urlSerializer.parse(redirect)` (the `redirectTo` target string, not the in-flight `UrlTree`), and any `:param` segments from the in-flight URL are reused by reference (`UrlSegment` instances already created by `url_tree.ts`).
   - In `create_url_tree.ts`, `createSegmentGroupFromRoute` builds outlet maps from `ActivatedRouteSnapshot.children`, which only exist for outlets that matched a configured `Route` in the application's `Routes` config.
   - Making `setUrlDerivedKey` a file-local helper in `url_tree.ts` avoids exporting a V8-specific helper from `utils/collection.ts`.

2. **In `url_tree.ts`, drop `parseQueryParam` and simplify `setUrlDerivedKey`:**
   - `UrlTree.queryParams` only exists once per URL (1 object), so V8's per-object density check already bounds it and `parseQueryParam` doesn't need `setUrlDerivedKey`.
   - We don't need the canonical uint32 round-trip check (`Number(key) >>> 0`, `index !== 0xffffffff`, `String(index) === key`)—transitioning on `Number(key) >= 32` is sufficient and harmless for strings like `'032'` or `'1e3'`:

   ```ts
   const SLOW_ELEMENTS_SENTINEL = 0x40000000;

   function setUrlDerivedKey<T>(target: {[key: string]: T}, key: string, value: T): void {
     if (Number(key) >= 32 && !Object.hasOwn(target, SLOW_ELEMENTS_SENTINEL)) {
       target[SLOW_ELEMENTS_SENTINEL] = value;
       delete target[SLOW_ELEMENTS_SENTINEL];
     }
     target[key] = value;
   }
   ```

3. **Trim the unit tests in `url_serializer.spec.ts`:**
   - Instead of 9 separate `it()` blocks testing JS property descriptors, `delete`, `__proto__`, and uint32 boundary values, please condense this to 1–2 focused tests verifying that numeric matrix parameters (including `1073741824`) parse and serialize properly without leaving the sentinel key behind.

**Review by atscott — APPROVED** [review]

**Comment by SkyZeroZx** [comment]
@atscott Can this be considered a security issue due to the evidence in #70716 ? or  It's only a harderning

**Comment by atscott** [comment]
Will discuss with @alan-agius4 and team

**Comment by thePunderWoman** [comment]
This PR was merged into the repository. The changes were merged into the following branches:

- main: 37196053d5e8884097cbf71f59734748e7b0c7b3
- 22.1.x: 2cab6f59649af47dd86e2fc7dd0dd269209c58bc
- 22.2.x: 5af61216eab8bf4a6697a1d79bda3d857c06f89d

**Comment by SkyZeroZx** [comment]
> Will discuss with @alan-agius4 and team

@atscott Sorry to bother you again. I'd like to know the team's thoughts on this issue.

I've been digging more into the issue and tested it with a similar configuration like AWS ELB ( 16kb ) and Node.js  (which doesn't seem to have a default limit on URI size)

I found the next results.

`a;990;2522` parses into `{ "990": "", "2522": "" }`. V8 reads those keys as indices and allocates a dense elements store up to the largest one, so that single object lands at 30,485 bytes. With `a;aaa;bbbb` it's 68.

Last runs, Angular 22.1.4 on Node 24, three fresh runs per row:

| Via             | V8 old space | URI size    | Requests to OOM |
| --------------- | -----------: | ----------: | --------------: |
| Nginx 16k (ALB) |      256 MiB |    16,368 B |               6 |
| Nginx 16k (ALB) |      512 MiB |    16,368 B |              12 |
| Node, direct    |      512 MiB | 1,048,531 B |               1 |
| Node, direct    |     1024 MiB | 1,048,531 B |               1 |
| Node, direct    |     2048 MiB | 2,097,062 B |               1 |


See the POC for more details https://github.com/SkyZeroZx/angular-router-ssr-oom-lab

I think this it can be considerate a security issue due to the results

**Comment by atscott** [comment]
Yes, can you please create a PR for the LTS branch? (20.3.x)

**Comment by SkyZeroZx** [comment]
> Yes, can you please create a PR for the LTS branch? (20.3.x)

Sure I will do it
